### PR TITLE
sql: Cascading deletes fail when deleting thousands of rows across interleaved tables

### LINKED ISSUE #44158: Cascading deletes fail on interleaves tables with thousands of rows in child tables
**Describe the problem**

When attempting to cascade delete across interleaved tables, when thousands of rows will be deleted from child tables, get the following error:

```
root@localhost:7101/raphael> delete from viewable where id = 80004486;

pq: internal error: key did not match descriptor
DETAIL: stack trace:
/go/src/github.com/cockroachdb/cockroach/pkg/sql/delete_range.go:251: processResults()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/delete_range.go:189: startExec()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:352: func2()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:141: func1()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:671: visitInternal()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:108: visit()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:72: walkPlan()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:355: startExec()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:117: Start()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:724: Run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:375: Run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:372: Run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:997: PlanAndRun()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:825: execWithDistSQLEngine()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:717: dispatchToExecutionEngine()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:416: execStmtInOpenState()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:98: execStmt()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1243: execCmd()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1172: run()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:444: ServeConn()
/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584: func1()
/usr/local/go/src/runtime/asm_amd64.s:1337: goexit()

HINT: You have encountered an unexpected error.
```

I get a similar error in the logs of the gateway node. I turned on session tracing, and combined with above log, and confirmed that

- `deleteRangeNode.startExec()` sets an upper bound of the number of rows to delete per each batch (600)
- first batch executes successfully
- second batch fails in `deleteRangeNode.processResults()`

When the second batch starts, the deleting is already traversing down into the interleaved tables. Once the batch returns, successfully, from the lease holder for the range, the gateway then processes the result of the batch, attempting to count the rows. This function also checks that index key is valid ([`fetcher.ReadIndexKey()`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/delete_range.go#L179)), and this is where we fail. In our delete statement, we only identify the parent-level table (`viewable`), and thus `fetcher.Tables` only has one entry in it's slice. We then call [`sqlbase.DecodeIndexKeyWithoutTableIDIndexIDPrefix`](https://github.com/cockroachdb/cockroach/blob/9bcde7f7f10b796fe88a668413b40516b39cbfca/pkg/sql/row/fetcher.go#L740), but we [strip out the *parent* table's id and primary index Id](https://github.com/cockroachdb/cockroach/blob/9bcde7f7f10b796fe88a668413b40516b39cbfca/pkg/sql/row/fetcher.go#L752). I believe this is what causes `DecodeIndexKeyWithoutTableIDIndexIDPrefix()` to get confused and return an error.

I have a patch that I'll open a PR for shortly.

**To Reproduce**

Create hierarchy of interleaves tables (parent-> child-> grandchildren), using `ON CASCADE DELETE`. Insert thousands of rows per each parent row into the children/grandchildren, then try to delete a row from the parent.

**Expected behavior**
Cascading deletes should successfully delete rows form all the interleaved child tables.

**Environment:**
 - CockroachDB version 19.2.2
 - Server OS: Ubuntu 18.04

### PR DESCRIPTION
**Pull request by jasobrown**
Successive batches delete_range fail to decode index key properly.

Fixes #44158

Release note (bug fix): Fix bug when cascade deleting thousands of rows
across interleaved tables.

### CODE REVIEW

**Comment by rohany** [comment]
Hey jason, thanks for the patch! Your diagnosis of the problem looks correct, but I'm not sure if this is a correct solution.

I'm worried about the following case:
3 tables, t1, t2, t3. where t1 is interleaved by t2, and t2 is interleaved by t3.

A batch ends with prev = a key in t3. In the start of the next batch, we end up doing a delete on rows from t2, so we will see a key from t2, but since a key from t3 is not a prefix of it, we will do the `ReadIndexKey` check and still error. 

I think the correct solution instead would be just add all interleaved children of the table we are deleting to the args of the fetcher -- that way no matter where the batches are cut between the rows, `ReadIndexKey` is not going to falsely error.

I'd also add a logic test for this in maybe `interleave` to avoid regressing in the future. A compact repro is here:
```
CREATE TABLE t (x INT PRIMARY KEY);
CREATE TABLE t2 (x INT, y INT, PRIMARY KEY (x, y)) INTERLEAVE IN PARENT t(x);
ALTER TABLE t2 ADD CONSTRAINT fk FOREIGN KEY (x) REFERENCES t(x) ON DELETE CASCADE;
INSERT INTO t VALUES (1);
INSERT INTO t2 (SELECT 1, id FROM generate_series(1, 10000) AS id);
DELETE FROM t WHERE x = 1
```

**Comment by jasobrown** [comment]
Thanks for taking a quick look, @rohany!

> A batch ends with prev = a key in t3. In the start of the next batch, we end up doing a delete on rows from t2, so we will see a key from t2, but since a key from t3 is not a prefix of it, we will do the ReadIndexKey check and still error.

Hmm, i thought about that, as well, but clearly didn't account for that in the patch. 

> I think the correct solution instead would be just add all interleaved children of the table we are deleting to the args of the fetcher 

Agreed, and that was actually my first thought when approaching this, but I was unsure about it. I'll go ahead and implement that now that someone with more database-wide context advises it ;) I'll expand the interleave tests, as well.

**Comment by jasobrown** [comment]
@rohany RAFL

As you suggested, I switched to passing in the interleaved tables to the `Fetcher`, as well as adding a logictest to `interleaved`.

File: pkg/sql/delete_range.go [hunk 1]
```diff
@@ -119,10 +123,18 @@ func maybeCreateDeleteFastNode(
 		return nil, false
 	}
 
+	i := 0
```
Comment (rohany) [inline pkg/sql/delete_range.go]: [nit]: just use the iteration variable that the for loop provides (for i, tableEntry := range ... )

Comment (rohany) [reply pkg/sql/delete_range.go]: Or just allocate `interleavedDesc` with len = 0, capacity = len(fkTables) and append.

File: pkg/sql/opt_exec_factory.go [hunk 2]
```diff
@@ -1629,7 +1629,7 @@ func (ef *execFactory) ConstructDelete(
 
 	fastPathInterleaved := canDeleteFastInterleaved(tabDesc, fkTables)
 	if fastPathNode, ok := maybeCreateDeleteFastNode(
-		context.TODO(), input.(planNode), tabDesc, fastPathInterleaved, rowsNeeded); ok {
+		context.TODO(), input.(planNode), tabDesc, fkTables, fastPathInterleaved, rowsNeeded); ok {
```
Comment (rohany) [inline pkg/sql/opt_exec_factory.go]: Can you add a comment that `canDeleteFastInterleaved` actually modifies `fkTables` and adds all of the interleaved children? It wasn't clear to me that this was the case and I was confused why `fkTables` was the correct thing to pass in.

File: pkg/sql/row/fetcher.go [hunk 3]
```diff
@@ -742,6 +742,7 @@ func (rf *Fetcher) ReadIndexKey(
 ) (remaining []byte, ok bool, foundNull bool, err error) {
 	// If there is only one table to check keys for, there is no need
 	// to go through the equivalence signature checks.
+
```
Comment (rohany) [inline pkg/sql/row/fetcher.go]: [nit]: is this needed?

**Review by rohany — CHANGES_REQUESTED** [review]
Almost there! Just a few nits. Also, make sure to squish your PR into a single commit.

File: pkg/sql/opt_exec_factory.go (hunk 2 above)
Comment (rohany) [reply pkg/sql/opt_exec_factory.go]: can you update this to not use `context.TODO()` and instead use the context at the top of the function?

File: pkg/sql/row/fetcher.go (hunk 3 above)
Comment (jasobrown) [reply pkg/sql/row/fetcher.go]: leftover from a whole lot of debug print statements ;) removing.

File: pkg/sql/delete_range.go (hunk 1 above)
Comment (jasobrown) [reply pkg/sql/delete_range.go]: `fkTables` is a `map`, so i can't use a range iterator and get the index (like when iterating over a slice). Either way, your second proposal is best, I think.

File: pkg/sql/opt_exec_factory.go (hunk 2 above)
Comment (jasobrown) [reply pkg/sql/opt_exec_factory.go]: removed `context.TODO()`.

Looking at `canDeleteFastInterleaved()`, I'm not sure it modifies `fkTables`. We iterate over `fkTables`, for sure, but where does it modify?

**Comment by jasobrown** [comment]
RFAL

Addressed most of the nits, squashed, and pushed. I wasn't sure about the squashing or just pushing new commits, so thanks for clarifying.

File: pkg/sql/delete_range.go (hunk 1 above)
Comment (rohany) [reply pkg/sql/delete_range.go]: ack woops.

File: pkg/sql/opt_exec_factory.go (hunk 2 above)
Comment (rohany) [reply pkg/sql/opt_exec_factory.go]: Sorry, I read the code wrong. I thought the loop that `canDeleteFastInterleaved()` was modifying fkTables. 

After a second glance, I see why this is correct. By default, `fkTables` doesn't hold all of the interleaved children of the table. However, for us to actually enter the fast path for deletion, all the interleaved children have to be in `fkTables`, so it is correct for `fkTables` to be used as the set of tables given to the fetcher. Does my reasoning match what you intended?

**Review by jasobrown — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44159)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jasobrown and @rohany)

___
*[pkg/sql/opt_exec_factory.go, line 1632 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44159#-LzSg7jJxuzOlrrcM031-r2-1632:-LzbSK8_59sdwEI2NplV:b-8ygxdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2e6dd6afa310cf021ba47187e5479ef16a819ff/pkg/sql/opt_exec_factory.go#L1632)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Sorry, I read the code wrong. I thought the loop that `canDeleteFastInterleaved()` was modifying fkTables. 

After a second glance, I see why this is correct. By default, `fkTables` doesn't hold all of the interleaved children of the table. However, for us to actually enter the fast path for deletion, all the interleaved children have to be in `fkTables`, so it is correct for `fkTables` to be used as the set of tables given to the fetcher. Does my reasoning match what you intended?
</blockquote></details>

My understanding is that `fktables` *will* have all of the interleaved children, as per the implementation of `row.MakeFkMetadata()` (although I haven't dug too deeply into `tableLookupQueue` and it's implementation as it's used in `row.MakeFkMetadata()`). I also rely on `canDeleteFastInterleaved()` to do most of the correctnes checking since it ensures, amongst other things, that the child table's forign key is `ON DELETE CASCADE` (otherwise we wouldn't be able to fast path delete).

In other words, yes your reasoning does matches with what I'm trying to do :)

**Comment by rohany** [comment]
Could you drop a comment or two about this discussion in the code? It's pretty tricky, so we should have this documented. Other than this, this looks good!

**Review by rohany — APPROVED** [review]

**Comment by rohany** [comment]
Nice, thanks for your work!

bors r=rohany

**Comment by jasobrown** [comment]
@rohany woo-hoo! Thanks for help and guidance.
